numpy/random/mtrand/mtrand.pyx

```diff
@@ -4652,33 +4652,35 @@ cdef class RandomState:
                [0, 1, 2]])
 
         """
-        cdef npy_intp i, j
-
-        i = len(x) - 1
+        cdef:
+            npy_intp[::1] idxs
+            npy_intp i, j, n = len(x)
 
         # Logic adapted from random.shuffle()
-        if isinstance(x, np.ndarray) and \
-           (x.ndim > 1 or x.dtype.fields is not None):
-            # For a multi-dimensional ndarray, indexing returns a view onto
-            # each row. So we can't just use ordinary assignment to swap the
-            # rows; we need a bounce buffer.
-            buf = np.empty_like(x[0])
-            with self.lock:
-                while i > 0:
-                    j = rk_interval(i, self.internal_state)
-                    buf[...] = x[j]
-                    x[j] = x[i]
-                    x[i] = buf
-                    i = i - 1
+        if isinstance(x, np.ndarray):
+            if (x.ndim == 1 and
+                x.dtype.itemsize == np.dtype(np.intp).itemsize and
+                not isinstance(x, np.ma.MaskedArray)):
```

Comment:
I guess to be safer we might as well make it `type(x) is np.ndarray` to disallow all subclasses (since there's no guarantee that `MaskedArray` is the only weird subclass)